File: neutron.py
'''
 @file      neutron.py
 @brief     Contains the Neutron class
 @author    Geoffrey Gunow
 @date      October 7, 2015
'''

import numpy as np
from math import *
from copy import deepcopy as copy

'''
 @class Neutron neutron.py "neutron.py"
 @brief Contains neutron position and direction information
'''
class Neutron():
    def __init__(self, position, theta, phi, group):
        
        self._xyz = np.array(copy(position))
        self._alive = True
        self._group = group
        self.set_direction(theta, phi)
        self.association = {'x': 0, 'y':1, 'z':2}

    @property
    def x(self):
        return self._xyz[0]

    @property
    def y(self):
        return self._xyz[1]

    @property
    def z(self):
        return self._xyz[2]

    @property
    def alive(self):
        return self._alive

    @property
    def group(self):
        return self._group
    '''
     @brief move a neutron a given distance along its direction of travel
    '''
    def move(self, distance):
        self._xyz += self._direction * distance 

    '''
     @brief reverse the direction of travel along a given axis
    '''
    def reflect(self, axis):
        self._direction[self.association[axis]] *= -1

    '''
     @brief set the neutron's direction of travel given its angle in sphereical
            coordinates
    '''
    def set_direction(self, theta, phi):
        self._direction = np.array([sin(theta)*cos(phi), sin(theta)*sin(phi),
            cos(theta)])

    '''
     @brief set the neuton's position
    '''
    def set_position(self, var, value):
        self._xyz[self.association[var]] = value
    
    '''
     @brief set the neutron's cell
    '''
    def set_cell(self, cell_number):
        self._cell = cell_number

    '''
     @brief kill the neutron
    '''
    def kill(self):
        self._alive = False

    '''
     @brief return the cell of the neutron
    '''
    def get_cell(self):
        return self._cell
    
    '''
     @brief return the position of the neutron along a given axis
    '''
    def get_position(self, var):
        return self._xyz[self.association[var]]

    '''
     @brief return the position of the neutron in all three dimensions
    '''
    def get_position_vector(self):
        return copy(self._xyz)
    
    '''
     @brief return the neutron's direction of travel along a given axis
    '''
    def get_direction(self, var):
        return self._direction[self.association[var]]
        
    '''
     @brief return the neutron's direction of travel in all three dimensions
    '''
    def get_direction_vector(self):
        return copy(self._direction)

    '''
     @brief return the neutron's distance from the origin
    '''
    def get_distance_from_origin(self):
        return sqrt(self.x**2 + self.y**2 + self.z**2)

    '''
     @brief return the neutron's distance from a given coordinate
    '''
    def get_distance(self, coord):
        return sqrt((self.x - coord[0])**2 + (self.y - coord[1])**2 + \
                (self.z - coord[2])**2)

    '''
     @brief return's data about the neutron
    '''
    def __repr__(self):
        string = "<Neutron at position " + str(self._xyz) + " traveling in " \
                + "direction " + str(self._direction) + ">"
        return string

    '''
     @brief set the neutron's energy group
    '''
    def set_group(self, new_group):
        self._group = new_group

File: neutron.cpp
/** 
 @file      neutron.cpp
 @brief     contains functions for the Neutron class
 @author    Luke Eure
 @date      January 8 2016
*/

#include "neutron.h"


/**
 @brief constructor for Neutron class
 @param position a vector containing the starting position of the neutron
 @param theta the polar angle of the neutron's initial direction
 @param phi the azimuthal angle of the neutron's initial direction
*/
Neutron::Neutron(std::vector <double> &position, double theta, double phi) {
    _xyz = position;
    _neutron_alive = true;
    _neutron_direction.resize(3);
    setDirection(theta, phi);
}

/**
 @brief deconstructor for Neutron class
*/
Neutron::~Neutron() {}

/**
 @brief tells if the neutron is alive
 @return a bool (true=neutron alive. false=neutron dead)
*/
bool Neutron::alive(){
    return _neutron_alive;
}

/**
 @brief returns the energy group of the neutron
 @return an int: the energy group of the neutron
*/
int Neutron::getGroup() {
    return _neutron_group;
}

/**
 @brief moves the neutron a given distance
 @param distance the distance the neutron should be moved
*/
void Neutron::move(double distance) {
    for (int i=0; i<3; ++i) {
        _xyz[i] += _neutron_direction[i]*distance;
    }
}

/**
 @brief moves the neutron a given distance
 @param distance the distance the neutron should be moved
*/
void Neutron::reflect(int axis) {
    _neutron_direction[axis] *= -1;
}

/**
 @brief sets the neutron's direction of travel
 @param theta the polar angle
 @param phi the azimuthal angle
*/
void Neutron::setDirection(double theta, double phi) {
    _neutron_direction[0] = sin(theta)*cos(phi);
    _neutron_direction[1] = sin(theta)*sin(phi);
    _neutron_direction[2] = cos(theta);
}

/**
 @brief sets the neutron's position along an axis
 @param axis the axis along which the position will be set
 @param value the value to which the position will be set
*/
void Neutron::setPosition(int axis, double value) {
    _xyz[axis] = value;
}

/**
 @brief sets the cell of the neutron
 @param cell_number the cell to which the nuetron will be set
*/
void Neutron::setCell(std::vector <int> &cell_number) {
    _neutron_cell = cell_number;
}

/**
 @brief kills the neutron
*/
void Neutron::kill() {
    _neutron_alive = false;
}

/**
 @brief returns the neutron's cell
 @return the cell in which the neutron resides
*/
std::vector <int> Neutron::getCell() {
    return _neutron_cell;
}

/**
 @brief gets the position of the neutron along a certain axis
 @param axis an int containing the axis along which the position will be
        returned
 @return a double denoting the neutron's position along axis
*/
double Neutron::getPosition(int axis) {
    return _xyz[axis];
}

/**
 @brief gets the position vector of the neutron
 @return a vector containing the neutron's position
*/
std::vector <double> Neutron::getPositionVector() {
    return _xyz;
}

/**
 @brief gets the direction of the neutron along a certain axis
 @param axis an int containing the axis along which the direction will be
        returned
 @return a double denoting the neutron's direction along axis
*/
double Neutron::getDirection(int axis) {
    return _neutron_direction[axis];
}

/**
 @brief gets the direction vector of the neutron
 @return a vector containing the neutron's direction
*/
std::vector <double> Neutron::getDirectionVector() {
    return _neutron_direction;
}

/**
 @brief gest the neutron's distance from a given point
 @param coord a vector denoting the point to find the neutron's distance from
 @return the neutron's distance from that point
*/
double Neutron::getDistance(std::vector <double> &coord) {
    return sqrt(pow(getPosition(0)-coord[0], 2.0)
            + pow(getPosition(1)-coord[1], 2.0)
            + pow(getPosition(2)-coord[2], 2.0));
}

/**
 @brief set the neutron's group
 @param new_group the new energy group of the neutron
*/
void Neutron::setGroup(int new_group) {
    _neutron_group = new_group;
}
